BUG: CustomBusinessMonthBegin(End) sometimes ignores extra offset (GH41356)

jackzyliu opened the pull request:
- [x] closes #41356 #14869
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

-----------------
**Problem description**: _CustomBusinessMonth does not apply extra offset when initially rolled month begin `new` is already a business day. This happens because we currently use CustomBusinessDay.rollforward (rollback) to both 1) roll to a business day and 2) apply the extra offset. 
**Fix**: Separating the logic to 1) roll to a business day and 2) apply the extra offset.

jackzyliu committed:
BUG: CustomBusinessMonthBegin(End) sometimes ignores extra offset (GH41356)

CustomBusinessMonthBegin(End) does not apply extra offset when the initially rolled month begin (end) is already a business day.

jackzyliu commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -3350,7 +3350,13 @@ cdef class _CustomBusinessMonth(BusinessMixin):
+        else:
+            cbday_kwds = self.kwds
+
+        cbday = CustomBusinessDay(n=1, normalize=False, **cbday_kwds)
```
Changing from n=self.n to n=1 does not have any effect on the functionality. I just personally thought having n=1 is 1) more consistent with MonthBegin(n=1) and MonthEnd(n=1) below; and 2) slightly more indicative of the role of CustomBusinessDay, which is simply to apply a 1-day rollforward/rollback. Happy to change it back.

jbrockmendel commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -971,6 +971,7 @@ Other
 - Bug in :func:`pandas.util.show_versions` where console JSON output was not proper JSON (:issue:`39701`)
 - Bug in :meth:`DataFrame.convert_dtypes` incorrectly raised ValueError when called on an empty DataFrame (:issue:`40393`)
 - Bug in :meth:`DataFrame.clip` not interpreting missing values as no threshold (:issue:`40420`)
+- Bug in :meth:`_CustomBusinessMonth.apply` not applying extra offset when initially rolled date `new` is already a business day (:issue:`41356`)
```
_CustomBusinessMonth is not user-facing.  can this be phrased in terms of something users will directly see?

jackzyliu replied:
Thanks for the feedback! I just re-phased the entry. Is it more along the line of what you were thinking?

jreback commented on pandas/tests/tseries/offsets/test_month.py:
```diff
@@ -204,6 +205,56 @@ def test_datetimeindex(self):
+                datetime(2021, 3, 2): datetime(2021, 3, 30) + timedelta(days=1),
+            },
+        ),
+    ]
```
can you just put these directly in the paremeterize?

jackzyliu replied:
Thanks for the review! Will do. 

As a side note, the existing [tests](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/tseries/offsets/test_month.py#L125-L167) for `.apply` also put the case definition outside of `parametrize` . Happy to quickly change these as well if you would like.

jreback commented on pandas/tests/tseries/offsets/test_month.py:
```diff
@@ -341,6 +392,58 @@ def test_datetimeindex(self):
+        ),
+    ]
+
+    @pytest.mark.parametrize("case", apply_with_extra_offset_cases)
```
same

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -3350,7 +3350,13 @@ cdef class _CustomBusinessMonth(BusinessMixin):
         Define default roll function to be called in apply method.
         """
-        cbday = CustomBusinessDay(n=self.n, normalize=False, **self.kwds)
+        if self.offset:
```
isn't offset *always* defined?

jackzyliu replied:
Yes, the purpose of the check here is to make sure we only pass in a zero `self.offset` into `CustomBusinessDay` (and that we do not do unnecessary copying of the kwd dict).

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -3350,7 +3350,13 @@ cdef class _CustomBusinessMonth(BusinessMixin):
+        else:
+            cbday_kwds = self.kwds
+
+        cbday = CustomBusinessDay(n=1, normalize=False, **cbday_kwds)
```
can you not just pass kwds & offset here?

jackzyliu replied:
Likely I didn't fully understanding your suggestion here. Do you by any chance mean passing `**self.kwds` as before? So the idea here is that if _self.offset is not timedelta(0)_, then we always use `CustomBusinessDay(n=1, normalize=False, **offset=timedelta(0)**, ...)` to roll to a business day first, and then separately add the self.offset afterwards. 

The reason is that `.rollforward` (and `.rollback`) only calls `.apply` when if not `.is_on_offset(dt) (code [here](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/offsets.pyx#L590-L616) and [here](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/offsets.pyx#L3227-L3255)), which I think can be illustrated well with the following example:
`
In[33]: import pandas as pd
   ...: print(pd.__version__)
1.3.0.dev0+1628.g32ce31be89
In[34]: cbday_plus_5d = pd.offsets.CustomBusinessDay(n=1, offset=timedelta(days=5))
   ...: # define a CBDay with an extra offset of 5 days
In[35]: print(cbday_plus_5d.rollforward(pd.Timestamp('2021-05-02')))  # expected 2021-05-08
   ...: # 2021-05-02 not a business day, so 'rollforward' makes it 2021-05-03 + extra 5 days
2021-05-08 00:00:00
In[36]: print(cbday_plus_5d.rollforward(pd.Timestamp('2021-05-03')))  # expected 2021-05-08
   ...: # 2021-05-03 is ALREADY a business day, so it doesn't get rolled at all.
2021-05-03 00:00:00
`
So instead, we enforce `offset=timedelta(0)` while constructing `CustomBusinessDay` as follows, which is why I am first modifying the 'offset' parameter:
`
In[42]: cbday_no_offset = pd.offsets.CustomBusinessDay(n=1, **offset=timedelta(0)**)  
In[43]: print(cbday_no_offset.rollforward(pd.Timestamp('2021-05-03')) + timedelta(days=5))  # expected 2021-05-08
2021-05-08 00:00:00
`

Please let me know if I completely missed your question. Would love to hear your thoughts. Thanks!

jbrockmendel replied:
I think what jreback is suggesting is:

```
if self.offset:
    cbday_kwds = self.kwds.copy()
    cbday_kwds['offset'] = timedelta(0)
else:
    cbday_kwds = self.kwds
```

should instead just be:

```
cbday_kwds = self.kwds.copy()
cbday_kwds['offset'] = timedelta(0)
```

jackzyliu replied:
Ah, I see. I added the check here to prevent unnecessary data copying, because I am guessing `offset=timedelta(0)` is the more common use case and works as expected. What do you think? Definitely happy to change it accordingly.

jackzyliu replied:
Bumping this thread. Let me know what you think about 1) always making a new copy `cbday_kwds = self.kwds.copy()` vs. 2) only copying when `self.offset` is not `timedelta(0)`.

As a refresher for your convenience, the idea is that we want to first use `CustomBusinessDay(n=1, normalize=False, offset=timedelta(0), ...)` (instead of any non-zero offset) to roll to a business day first, and then separately apply the `self.offset` afterwards in `.apply`.

jbrockmendel replied:
im fine with this either way

jbrockmendel commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -974,6 +974,7 @@ Other
 - Bug in :meth:`DataFrame.convert_dtypes` incorrectly raised ValueError when called on an empty DataFrame (:issue:`40393`)
 - Bug in :meth:`DataFrame.clip` not interpreting missing values as no threshold (:issue:`40420`)
 - Bug in :class:`Series` backed by :class:`DatetimeArray` or :class:`TimedeltaArray` sometimes failing to set the array's ``freq`` to ``None`` (:issue:`41425`)
+- Bug in :meth:`CustomBusinessMonthBegin` (:meth:`CustomBusinessMonthEnd`) not applying the extra ``offset`` parameter when beginning (end) of the target month is already a business day (:issue:`41356`)
```
you've given the class but not the method.  Is it e.g. `CustomBusinessMonthBegin.__add__`?

jackzyliu replied:
Oops, my apologies. It should be corrected now.

simonjayhawkins commented:
@jackzyliu can you merge master to resolve conflicts

jackzyliu commented:
Hi @simonjayhawkins, I just rebased after master, but some checks started failing and I can't replicate the failed tests offline. Examples: 
- `pandas/tests/tseries/offsets/test_offsets_properties.py::test_on_offset_implementations` with actions-38.yaml
- `pandas/tests/arithmetic/test_datetime64.py::TestDatetimeIndexArithmetic::test_sub_dti_dti` with actions-39-numpydev.yaml
 
These checks were successful before rebasing without any code change. What could be the issue here?

simonjayhawkins commented:
The test_on_offset_implementations failure is a hypothesis failure so that maybe the reason it didn't fail before. may need to investigate that one further. The numpydev failures are a known issue.

jackzyliu commented:
Re-rebased after master. @simonjayhawkins 

> The test_on_offset_implementations failure is a hypothesis failure so that maybe the reason it didn't fail before. may need to investigate that one further. The numpydev failures are a known issue.

Previously, `pandas/tests/tseries/offsets/test_offsets_properties.py::test_on_offset_implementations` failed with the following pair of randomly generated test inputs,:
```
dt = datetime.datetime(1900, 1, 1, 0, 0, tzinfo=pytz.timezone('Africa/Kinshasa'))
offset = pd.offsets.MonthBegin(66)
``` 
Not sure if this is due to some issue within `tz_localize_to_utc` or just an unlucky test case. I will create a separate issue and do a bit more investigation.

jackzyliu committed:
whatsnew entry (GH41356)

jackzyliu commented:
@jreback @simonjayhawkins 

Bumping up this PR. Since we just released `v1.3.0`, I have moved the whatsnew entry to `1.3.1` for now. Happy to move it to `1.4.0` instead. 

The failed test seem to come from master?

Please let me know of further comments. Thanks.

jackzyliu commented:
@jreback @jbrockmendel @simonjayhawkins 

- Moved whatsnew entry to `1.4.0`
- Rebased after master
- All checks green

Please let me know of further comments. Thanks!

jackzyliu commented:
Resolved conflict with master. Green.

jreback commented:
thanks @jackzyliu, sorry about the wait, lots of PRs :->

keep em coming!
